### ISSUE TITLE
Padding question

### ISSUE THREAD

**Issue by ogozman** [seladb/PcapPlusPlus#1012]
Hello, I seem to have issue with the location of packet trailer layer in payload. When I get UDPLayer, the actual data in payload that goes after header is offset by the number of bytes that is always equal to the length of packet trailer layer that is present in both datagram types (there are 2 types of PDUs and both have the same offset). The question is, is it a bug or is it a consistent behavior that is supposed to happen? I thought that padding should go after the data, not between payload and header. I am using IPv4 reassembly in this case. I have no issue with short PDU where no reassembly is needed. Thank you!

**Comment by seladb** [comment]
Hi @ogozman ! I'm not sure I understand the issue. Can you provide 2 pcap files that reproduce the issue?
- One with the fragments before running IP reassembly
- Another with the output of IP reassembly and point to where the problem is?

**Comment by ogozman** [comment]
Thank you for the fast response, @seladb. I will make a setup and upload the pcap file soon.

**Comment by ogozman** [comment]
Hey, as discussed, here is a zip file with 2 pcap files, one is defragmented packet (ipv4reassembly) and the other one is fragmented packet and (if I understand correctly) packet reassembled by OS. I have actually a question in regards to my assumption, does PCAP++ also capture reassembled packets? Currently I am gathering IPv4 packets and reassembling myself, but the sender actually transmits UDP packets. Can I just capture those and avoid reassembly? 
If you compare OS reassembled package and my reassembled package, you can clearly see a bunch of 0s (14 to be exact) that are only in manually reassembled package.

**Comment by seladb** [comment]
Thanks for sharing your pcap files @ogozman , I'll take a look and get back to you

**Comment by seladb** [comment]
I think I found the issue. Packet no. 8 in your file, which is IP fragment no. 1 has a packet trailer of 12 bytes. The IP reassembly logic uses [`getLayerPayloadSize()`](https://github.com/seladb/PcapPlusPlus/blob/f6baff3c0eac1b66cc96861ed2be3e6a9c43f8b8/Packet%2B%2B/header/Layer.h#L112) to calculate the IP layer payload size which is actually the total size of the IP payload minus the IP header size:

https://github.com/seladb/PcapPlusPlus/blob/f6baff3c0eac1b66cc96861ed2be3e6a9c43f8b8/Packet%2B%2B/src/IPReassembly.cpp#L129

https://github.com/seladb/PcapPlusPlus/blob/f6baff3c0eac1b66cc96861ed2be3e6a9c43f8b8/Packet%2B%2B/src/IPReassembly.cpp#L215

That's why the packet trailer of fragment 1 is included in the defragmented packet.

Instead of using `getLayerPayloadSize()` we should probably calculate the IPv4/IPv6 payload length from the IP header (i.e `Total Length` for IPv4 and `Length` for IPv6).

Would you consider making this fix for this?

You can add a test using your example pcap files.

**Comment by ogozman** [comment]
I see. I am a bit new to all this networking stuff, but shouldn't we use then, with ipv4, at least, something like `ipv4Layer->getIPv4Header()->totalLength - ipv4Layer->getHeaderLen()`? Since we need to get the actual payload without padding, so we would have total valid data length and subtract header length.

**Comment by ogozman** [comment]
I have tried to make a change but it did not lead to any changes in the outcome after reassembly. In fact, I have checked the return value of `m_IPLayer->getLayerPayloadSize()` in case of padded packet and it was correct all along, 8 bytes, not 22 bytes. The issue must lie somewhere else.

**Comment by seladb** [comment]
Thank you for debugging the issue @ogozman , I think you found the bug!

Doing `fragment->removeLayer(pcpp::PacketTrailer)` is probably not a good idea because `fragment` is a pointer to a packet that is part of the user input, hence we don't want to change it.

On the other hand, we do want to create a `RawPacket` that will contain all the layers until IP. Maybe we should create the `RawPacket` in a different way that will not copy the packet trailer data, for example:

```cpp
RawPacket* fragmentRawPacket = fragment->getRawPacket();
fragData->data = new RawPacket(fragmentRawPacket >getRawData(), fragWrapper->getIPLayerPayload() - fragmentRawPacket ->getRawData() + fragWrapper->getIPLayerPayloadSize(), fragmentRawPacket->getPacketTimeStamp(), fragmentRawPacket->getLinkLayerType());
```

I didn't actually run this code, but I hope you get the point: we take the fragment data from start until the end of the IP layer payload, hence avoiding everything beyond the IP layer including the packet trailer.

Please let me know if it makes sense.

**Comment by seladb** [comment]
Thank you @ogozman for opening this issue and help to debug it! 🙏 

Fixed in #1066 , closing the issue now
